env.bzl

```
@@ -23,18 +36,31 @@ def _clean(rctx, executable):
     if result.return_code:
         fail("Failure cleaning up.\nstdout: {}\nstderr: {}".format(result.stdout, result.stderr))
 
+def _create_empty_environment(rctx, executable, env_name):
+    rctx.report_progress("Creating empty conda environment, to be populated afterwards")
+    args = [rctx.path(executable), "create", "-y", "-p", "./{}".format(env_name)]
+    result = rctx.execute(args, quiet = rctx.attr.quiet, timeout = rctx.attr.timeout)
+    if result.return_code:
+        fail("Failure creating empty environment.\nstdout: {}\nstderr: {}".format(result.stdout, result.stderr))
+
+def _update_environment(rctx, executable, env_name, env_file):
+    rctx.report_progress("Updating empty conda environment to populate it")
+    env_file = str(rctx.path(rctx.attr.environment))
+    args = [rctx.path(executable), "env", "update", "-f", env_file, "-p", "./{}".format(env_name)]
+    result = rctx.execute(args, quiet = rctx.attr.quiet, timeout = rctx.attr.timeout)
+    if result.return_code:
+        fail("Failure updating environment.\nstdout: {}\nstderr: {}".format(result.stdout, result.stderr))
+
 # create new local conda environment from file
 def _create_environment(rctx, executable, env_name):
     rctx.report_progress("Creating conda environment")
 
+    executable = _user_chosen_executable(rctx)
```

Maybe let's resolve the path to the executable here and just pass a path to later steps:
```suggestion
    executable = str(rctx.path(_user_chosen_executable(rctx)))
```

---

Ok, I will resolve upon assignment. But the outer `str()` is not needed. This works:

    executable = rctx.path(_user_chosen_executable(rctx))
    ....
    rctx.execute(executable, ....)

Because rctx.execute expects a path as its first argument. See https://docs.bazel.build/versions/main/skylark/lib/repository_ctx.html#execute

---

Well, to be more precise, the docs just say it accepts a path as its first argument but do not specify that the path must be a string. Since `rctx.path()` returns a path, whether that is a string or a path object, I don't know. I expect this way to continue to work.

---

Oh ok, cool. If it works, it works 😄